cohorts/cohort.py

```diff
@@ -470,23 +489,117 @@ def load_variants(self, patients=None, filter_fn=None):
+        if filter_fn is None:
+            filter_fn_name = 'None'
+        else:
+            filter_fn_name = filter_fn.__name__
+        logger.debug('loading variants with filter_fn: {}'.format(filter_fn_name))
         patient_variants = {}
 
         for patient in self.iter_patients(patients):
-            variants = self._load_single_patient_variants(patient, filter_fn)
+            variants = self._load_single_patient_variants(patient, filter_fn, **kwargs)
             if variants is not None:
                 patient_variants[patient.id] = variants
         return patient_variants
+    
+    
+    def _hash_filter_fn(self, filter_fn, **kwargs):
+        if filter_fn is None:
+            return 'filter-none'
+        filter_fn_name = filter_fn.__name__
+        logger.debug('Computing hash for filter_fn: {} with kwargs {}'.format(filter_fn_name, str(dict(**kwargs))))
+        # function source code
+        fn_source = str(dill.source.getsource(filter_fn))
+        pickled_fn_source = pickle.dumps(fn_source)
+        # kwarg values
+        kw_dict = dict(**kwargs)
+        kw_hash = list()
+        if not kw_dict:
+            kw_hash = ['default']
+        else:
+            [kw_hash.append('{}-{}'.format(key, h)) for (key, h) in sorted(kw_dict.items())]
+        # closure vars
+        closure = 'null'
+        nonlocals = inspect.getclosurevars(filter_fn).nonlocals
+        for (key, val) in nonlocals.items():
+            if inspect.isfunction(val):
+                closure = '{}-{}'.format(self._hash_filter_fn(val).__name__, self._hash_filter_fn(val))
+        # construct final hashed_fn
+        hashed_fn = '.'.join(['-'.join([filter_fn.__name__,
+                                        int(hashlib.sha1(pickled_fn_source).hexdigest(), 16) % (10 ** 11)]),
+                              '.'.join(kw_hash),
+                              closure]
+                            )
+        return hashed_fn
+            
+        
+    
+    def _load_single_patient_variants(self, patient, filter_fn, use_cache=True, **kwargs):
+        if filter_fn is None:
+            filter_fn_name = 'None'
+        else:
+            filter_fn_name = filter_fn.__name__
+        logger.debug('loading variants for patient {} with filter_fn {}'.format(patient.id, filter_fn_name))
+        use_filtered_cache = use_cache
+        if sys.version_info < (3, 3):
+            logger.debug('... disabling filtered cache due to python version')
+            use_filtered_cache = False
+            
+        ## get cache name, if possible
+        if use_filtered_cache:
+            logger.debug('... identifying filtered-cache file name')
+            try:
+                ## try to load filtered variants from cache
+                filtered_cache_file_name = "%s-variants.%s.pkl" % (self.merge_type,
+                                                                          self._hash_filter_fn(filter_fn, **kwargs))
+            except:
```

Comment:
That's not equivalent logical structure - making it equivalent will end up with what (IMO) will end up being a more confusing set of nested Try & If statements.  You can think of the second `if` as a "if we are *still* using filtered cache".